src/clientrpc.cpp

```diff
@@ -94,11 +98,127 @@ CClientRpc::CClientRpc ( CClient* pClient, CRpcServer* pRpcServer, QObject* pare
+    /// @param {number} params.state - The recorder state
+    connect ( pClient, &CClient::RecorderStateReceived, [=] ( const ERecorderState newRecorderState ) {
+        pRpcServer->BroadcastNotification ( "jamulusclient/recorderState", QJsonObject{ { "state", newRecorderState } } );
+    } );
+
+    /// @rpc_method jamulus/pollServerList
+    /// @brief Request list of servers in a directory
+    /// @param {string} params.directory - socket address of directory to query, e.g. anygenre1.jamulus.io:22124.
+    /// @result {string} result - "ok" or "error" if bad arguments.
+    pRpcServer->HandleMethod ( "jamulusclient/pollServerList", [=] ( const QJsonObject& params, QJsonObject& response ) {
+        auto jsonDirectoryIp = params["directory"];
+        if ( !jsonDirectoryIp.isString() )
+        {
+            response["error"] = CRpcServer::CreateJsonRpcError ( CRpcServer::iErrInvalidParams, "Invalid params: directory is not a string" );
+            return;
+        }
+
+        CHostAddress haDirectoryAddress;
+        if ( NetworkUtil().ParseNetworkAddress ( jsonDirectoryIp.toString(), haDirectoryAddress, false ) )
+        {
+            // send the request for the server list
+            pClient->CreateCLReqServerListMes ( haDirectoryAddress );
+            response["result"] = "ok";
+        }
+        else
+        {
+            response["error"] =
+                CRpcServer::CreateJsonRpcError ( CRpcServer::iErrInvalidParams, "Invalid params: directory is not a valid socket address" );
+        }
+
+        response["result"] = "ok";
+    } );
+
+    /// @rpc_method jamulusclient/connect
+    /// @brief Disconnect client from server
+    /// @param {string} params.address - Server socket address (ip_addr:port).
+    /// @result {string} result - Always "ok".
+    pRpcServer->HandleMethod ( "jamulusclient/connect", [=] ( const QJsonObject& params, QJsonObject& response ) {
+        auto jsonAddr = params["address"];
+        if ( !jsonAddr.isString() )
+        {
+            response["error"] = CRpcServer::CreateJsonRpcError ( CRpcServer::iErrInvalidParams, "Invalid params: address is not a string" );
+            return;
+        }
+
+        if ( pClient->SetServerAddr ( jsonAddr.toString() ) )
+        {
+            if ( !pClient->IsRunning() )
+            {
+                pClient->Start();
+            }
+            response["result"] = "ok";
+        }
+        else
+        {
+            response["error"] = CRpcServer::CreateJsonRpcError ( 1, "Bad server address" );
+        }
+    } );
+
+    /// @rpc_method jamulusclient/disconnect
+    /// @brief Disconnect client from server
+    /// @param {object} params - No parameters (empty object).
+    /// @result {string} result - Always "ok".
+    pRpcServer->HandleMethod ( "jamulusclient/disconnect", [=] ( const QJsonObject& params, QJsonObject& response ) {
+        if ( pClient->IsRunning() )
+        {
+            pClient->Stop();
```

Comment:
I believe that this should call Disconnect() here instead https://github.com/jamulussoftware/jamulus/blob/e12cb71abc2de9477095ebfdfe62f31f15e09513/src/clientdlg.cpp#L1244 if the UI is running.

Reply:
Nothing should call Dialog methods.  Ever.  The Dialog UI might not exist if th RPC interface is in use.

Reply:
We need to update the UI somehow. -> refactoring needed.

Reply:
Yes.  If an a change to state happens to the Client (`CClient`), it should emit an event.  The UI (`CClientDlg`) can respond to that.  But the state should be in the Client itself, not the UI.  That way, it's accessible to both UI and RPC-I.